- Insert necessary library


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
import matplotlib.pyplot as plt
import requests
import pandas as pd
import numpy as np
import warnings
import json
import pandas as pd
warnings.filterwarnings('ignore')


**1. Calculate the percentage of Electric Vehicle in Sydney**

- Import the data

In [3]:
# Correct file path
file_path = '/content/drive/MyDrive/SIT764/Data/AAA-Postcode-Registrations-2024.xlsx'

# Read the Excel file (you can specify the sheet if needed)
ev_data_registration = pd.read_excel(file_path, sheet_name='Registration Numbers')

# Display the first 5 rows
ev_data_registration.head(5)


Postcode State    Fuel Type  Registrations as at 31 January 2021  \
0       800    NT          BEV                                    3   
1       800    NT  Hybrid/PHEV                                  131   
2       800    NT          ICE                                 6780   
3       810    NT          BEV                                    6   
4       810    NT  Hybrid/PHEV                                  288   

   Registrations as at 31 January 2022  Registrations as at 31 January 2023  \
0                                    3                                   12   
1                                  413                                  401   
2                                 7313                                 7088   
3                                   16                                   22   
4                                  375                                  467   

   Registrations as at 31 January 2024  
0                                   20  
1                                  435  
2                                 6574  
3                                   75  
4                                  592

In [4]:
# Fiter NSW only
vehicle_NSW_registration = ev_data_registration[ev_data_registration['State'] == 'NSW']
vehicle_NSW_registration.head(5)

Postcode State    Fuel Type  Registrations as at 31 January 2021  \
88      2000   NSW          BEV                                   76   
89      2000   NSW  Hybrid/PHEV                                  979   
90      2000   NSW          ICE                                18203   
91      2007   NSW          BEV                                    5   
92      2007   NSW  Hybrid/PHEV                                   65   

    Registrations as at 31 January 2022  Registrations as at 31 January 2023  \
88                                  159                                  272   
89                                 1218                                 1571   
90                                17924                                18017   
91                                   14                                   24   
92                                   81                                  100   

    Registrations as at 31 January 2024  
88                                  930  
89                                 1931  
90                                18390  
91                                   40  
92                                  110

In [5]:
# Check vehicle type
vehicle_NSW_registration['Fuel Type'].unique()

array(['BEV', 'Hybrid/PHEV', 'ICE'], dtype=object)

In [6]:
# Choose electric vehicle
ev_NSW_registration = vehicle_NSW_registration[vehicle_NSW_registration['Fuel Type'].isin(['Hybrid/PHEV', 'BEV'])]
ev_NSW_registration.head(5)

Postcode State    Fuel Type  Registrations as at 31 January 2021  \
88      2000   NSW          BEV                                   76   
89      2000   NSW  Hybrid/PHEV                                  979   
91      2007   NSW          BEV                                    5   
92      2007   NSW  Hybrid/PHEV                                   65   
94      2008   NSW          BEV                                    9   

    Registrations as at 31 January 2022  Registrations as at 31 January 2023  \
88                                  159                                  272   
89                                 1218                                 1571   
91                                   14                                   24   
92                                   81                                  100   
94                                   20                                   34   

    Registrations as at 31 January 2024  
88                                  930  
89                                 1931  
91                                   40  
92                                  110  
94                                   60

In [8]:

# Correct file path as a string
json_path = "/content/drive/MyDrive/SIT764/Data/vehicle_type.json"

# Open and load JSON file
with open(json_path, 'r') as file:
    raw_data = json.load(file)

# Convert the inner JSON strings into Python dictionaries
records = [json.loads(item["data"]) for item in raw_data["data"]["evIndexData"]]

# Create DataFrame
df_ev_index = pd.DataFrame(records)

# Display the first few rows
df_ev_index.head(5)


STATE  VEHICLE TYPE MANUFACTURER                    MODEL  \
0  NORTHERN TERRITORY     Small SUV       Suzuki           Suzuki S-Cross   
1     SOUTH AUSTRALIA     Large SUV      Hyundai  Hyundai Santa Fe Hybrid   
2            VICTORIA  People Mover          LDV                  LDV G10   
3          QUEENSLAND    Sports Car      Ferrari              Ferrari 488   
4     NEW SOUTH WALES    Sports Car       Nissan                 Nissan Z   

  FUEL TYPE  Q1 2022  Q2 2022  Q3 2022  Q4 2022  Q1 2023  Q2 2023  Q3 2023  \
0       ICE        1        0        0        0        3        1        2   
1    Hybrid        0        0        0       17       25       21       12   
2       ICE       51        8        2        0        0        0        0   
3       ICE        0        0        0        0        0        0        0   
4       ICE        0        0       20       24       32       36       57   

   Q4 2023  Q1 2024  Q2 2024  Q3 2024  Q4 2024  Q1 2025  
0        0        0        1        1        1        3  
1       24        7       52       84       79       83  
2        0        0        0        0        0        0  
3        0        0        0        0        0        0  
4       31       25       24       31       11       18

In [9]:
# Fiter NSW only
vehicle_NSW_data = df_ev_index[df_ev_index['STATE'] == 'NEW SOUTH WALES']
vehicle_NSW_data.head(5)


STATE VEHICLE TYPE MANUFACTURER                    MODEL  \
4   NEW SOUTH WALES   Sports Car       Nissan                 Nissan Z   
7   NEW SOUTH WALES    Small SUV       Subaru         Subaru XV Hybrid   
13  NEW SOUTH WALES   Medium SUV          BMW                   BMW X4   
15  NEW SOUTH WALES    Ute (2WD)   Mitsubishi  Mitsubishi Triton (2WD)   
16  NEW SOUTH WALES   Medium SUV       Toyota          Toyota RAV4 ICE   

   FUEL TYPE  Q1 2022  Q2 2022  Q3 2022  Q4 2022  Q1 2023  Q2 2023  Q3 2023  \
4        ICE        0        0       20       24       32       36       57   
7     Hybrid       33       38      105       59       44        1        0   
13       ICE       92       74       88       45       78       34       65   
15       ICE      296      249      292      191      206      222      124   
16       ICE      507     1041      302      739      428      339      287   

    Q4 2023  Q1 2024  Q2 2024  Q3 2024  Q4 2024  Q1 2025  
4        31       25       24       31       11       18  
7         0        0        0        0        0        0  
13       63       49       91       77       73       30  
15      289      245      199      215      216      154  
16      158      209      215      220       66       11

- Compute EV registered in NSW

In [10]:
# List of quarterly columns
quarter_columns = [col for col in vehicle_NSW_data.columns if col.startswith('Q')]

# Convert quarterly columns to yearly totals
# Create a new DataFrame with the same index
annual_sales = pd.DataFrame(index=vehicle_NSW_data.index)

# Loop through years and sum the quarters
for year in ['2022', '2023', '2024', '2025']:
    matching_quarters = [col for col in quarter_columns if year in col]
    annual_sales[year] = vehicle_NSW_data[matching_quarters].sum(axis=1)

# Combine with metadata columns
result_df = pd.concat([vehicle_NSW_data[['STATE', 'VEHICLE TYPE', 'MANUFACTURER', 'MODEL', 'FUEL TYPE']], annual_sales], axis=1)

result_df.head(5)

STATE VEHICLE TYPE MANUFACTURER                    MODEL  \
4   NEW SOUTH WALES   Sports Car       Nissan                 Nissan Z   
7   NEW SOUTH WALES    Small SUV       Subaru         Subaru XV Hybrid   
13  NEW SOUTH WALES   Medium SUV          BMW                   BMW X4   
15  NEW SOUTH WALES    Ute (2WD)   Mitsubishi  Mitsubishi Triton (2WD)   
16  NEW SOUTH WALES   Medium SUV       Toyota          Toyota RAV4 ICE   

   FUEL TYPE  2022  2023  2024  2025  
4        ICE    44   156    91    18  
7     Hybrid   235    45     0     0  
13       ICE   299   240   290    30  
15       ICE  1028   841   875   154  
16       ICE  2589  1212   710    11

In [12]:
# Check fuel type
result_df['FUEL TYPE'].unique()

array(['ICE', 'Hybrid', 'PHEV', 'BEV', 'HFCEV'], dtype=object)

In [14]:
# Filtered only electric vehicle type
ev_NSW_data = result_df[result_df['FUEL TYPE'].isin(['Hybrid', 'PHEV', 'BEV', 'HFCEV'])]
ev_NSW_data.head(5)

STATE VEHICLE TYPE MANUFACTURER                          MODEL  \
7   NEW SOUTH WALES    Small SUV       Subaru               Subaru XV Hybrid   
25  NEW SOUTH WALES    Large SUV       Toyota           Toyota Kluger Hybrid   
44  NEW SOUTH WALES    Small SUV   Mitsubishi  Mitsubishi Eclipse Cross PHEV   
45  NEW SOUTH WALES    Large SUV   Land Rover       Land Rover Defender PHEV   
47  NEW SOUTH WALES    Large Car          BMW              BMW 5 Series PHEV   

   FUEL TYPE  2022  2023  2024  2025  
7     Hybrid   235    45     0     0  
25    Hybrid  3023  2328  2544   666  
44      PHEV   252   366   656   152  
45      PHEV     0    11    59    16  
47      PHEV     1     2    16     6

In [22]:
# Calculate the sum of car registered in NSW
vehicle_NSW = result_df.drop(columns=['STATE','VEHICLE TYPE','MANUFACTURER','MODEL','FUEL TYPE'])
vehicle_total = vehicle_NSW.sum()
vehicle_total

2022    321895
2023    358471
2024    360754
2025     87480
dtype: int64

In [23]:
# Calculate the sum of EV registered in NSW
ev_NSW = ev_NSW_data.drop(columns=['STATE','VEHICLE TYPE','MANUFACTURER','MODEL','FUEL TYPE'])
ev_total = ev_NSW.sum()
ev_total

2022    41335
2023    65614
2024    91750
2025    24374
dtype: int64

In [24]:
# Calculate the percentage of EV in each year
ev_percentage = ev_total/vehicle_total *100
print(ev_percentage)

2022    12.841144
2023    18.303852
2024    25.432843
2025    27.862369
dtype: float64


In [26]:
# Calculate the mean
ev_average_percentage = ev_percentage.mean()
print(ev_average_percentage)

21.11005186566156


EV take 21.1 percent of everything


**2. Create the traffic volume datasets in Sydney**


In [27]:
# Read the charging station in Sydney
ev_charger_station_NSW = pd.read_excel('/content/drive/MyDrive/SIT764/Data/nsw_ev_chargers.xlsx')
ev_charger_station_NSW.head(5)

ObjId                  Station name  \
0      1                Curlew Retreat   
1      2              Donegal Farmstay   
2      3              Glenrock Gardens   
3      4  Lightning Ridge Bowling Club   
4      5                   Baths Motel   

                                     Station address Opening hours  \
0   399 Wylie Creek Rd, Maryland NSW 2372, Australia     Open 24/7   
1  103 Sandy Flat Road, Sandy Flat NSW 2372, Aust...           NaN   
2   84 Robinsons Ln, Tenterfield NSW 2372, Australia           NaN   
3  Lightning Ridge District Bowling Club, 1 Agate...     Open 24/7   
4  339 Warialda Street, Moree, New South Wales, A...           NaN   

                    Operator  Number of station  Number of plugs  \
0                        NaN                  1                1   
1          Tesla Destination                  2                4   
2          Tesla Destination                  2                2   
3          Tesla Destination                  2                2   
4  Tesla Destination, Everty                  1                2   

  Charger rating  Tesla  Type 2  J-1772   Latitude   Longitude  
0             kW      1       0       0 -28.558403  152.108476  
1          22 kW      2       2       0 -29.233601  152.022356  
2           6 kW      2       0       0 -29.002353  152.011511  
3           7 kW      2       0       0 -29.428298  147.977470  
4           7 kW      1       1       0 -29.475341  149.845634

In [5]:
# Read the traffic volume by hour in Sydney
vehicle_traffic_volume = pd.read_csv('/content/drive/MyDrive/SIT764/Data/road_traffic_counts_hourly_permanent.csv')

In [ ]:
vehicle_traffic_volume.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1038415 entries, 0 to 1038414
Data columns (total 41 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   cartodb_id              1038415 non-null  int64  
 1   the_geom                0 non-null        float64
 2   the_geom_webmercator    0 non-null        float64
 3   record_id               0 non-null        float64
 4   station_key             1038415 non-null  int64  
 5   traffic_direction_seq   1038415 non-null  int64  
 6   cardinal_direction_seq  1038415 non-null  int64  
 7   classification_seq      1038415 non-null  int64  
 8   date                    1038415 non-null  object 
 9   year                    1038415 non-null  int64  
 10  month                   1038415 non-null  int64  
 11  day_of_week             1038415 non-null  int64  
 12  public_holiday          1038415 non-null  bool   
 13  school_holiday          1038415 non-null  bool   
 14  da

In [ ]:
# Check for the missing values
vehicle_traffic_volume.isna().sum()


cartodb_id                      0
the_geom                  1038415
the_geom_webmercator      1038415
record_id                 1038415
station_key                     0
traffic_direction_seq           0
cardinal_direction_seq          0
classification_seq              0
date                            0
year                            0
month                           0
day_of_week                     0
public_holiday                  0
school_holiday                  0
daily_total                     0
hour_00                    124797
hour_01                    146679
hour_02                    150881
hour_03                    129513
hour_04                     86230
hour_05                     52469
hour_06                     41608
hour_07                     30342
hour_08                     29422
hour_09                     28107
hour_10                     26585
hour_11                     26715
hour_12                     26149
hour_13                     24767
hour_14                     26954
hour_15                     24726
hour_16                     29997
hour_17                     47008
hour_18                     51757
hour_19                     57925
hour_20                     64189
hour_21                     71388
hour_22                     85690
hour_23                    114133
md5                             0
updated_on                      0
dtype: int64

In [ ]:
# Check values in classification_seq
vehicle_traffic_volume['classification_seq'].unique()

array([2, 0, 3])

In [ ]:
# Filter light and heavy vehicle
all_vehicle_traffic_volume = vehicle_traffic_volume[vehicle_traffic_volume['classification_seq'].isin([2, 3])]
all_vehicle_traffic_volume.head(10)


cartodb_id  the_geom  the_geom_webmercator  record_id  station_key  \
0      3808503       NaN                   NaN        NaN     15286009   
1      3808504       NaN                   NaN        NaN     15168302   
3      3808506       NaN                   NaN        NaN        57159   
5      3808508       NaN                   NaN        NaN     21606004   
7      3808510       NaN                   NaN        NaN     99990004   
8      3808511       NaN                   NaN        NaN     18479662   
9      3808512       NaN                   NaN        NaN     21728002   
10     3808513       NaN                   NaN        NaN        58868   
11     3808514       NaN                   NaN        NaN        57186   
12     3808515       NaN                   NaN        NaN        57322   

    traffic_direction_seq  cardinal_direction_seq  classification_seq  \
0                       0                       5                   2   
1                       1                       7                   2   
3                       0                       7                   2   
5                       1                       7                   3   
7                       0                       3                   3   
8                       0                       5                   3   
9                       1                       1                   3   
10                      0                       5                   3   
11                      0                       1                   3   
12                      1                       1                   3   

                    date  year  ...  hour_16  hour_17  hour_18  hour_19  \
0   2020-04-01T00:00:00Z  2020  ...   1051.0    765.0    609.0    503.0   
1   2020-04-03T00:00:00Z  2020  ...    147.0    159.0    118.0     85.0   
3   2020-04-23T00:00:00Z  2020  ...   1061.0    856.0    409.0    202.0   
5   2020-04-18T00:00:00Z  2020  ...      4.0      5.0      6.0      1.0   
7   2020-04-24T00:00:00Z  2020  ...     17.0     18.0     21.0     16.0   
8   2020-04-07T00:00:00Z  2020  ...      7.0      3.0      1.0      8.0   
9   2020-04-18T00:00:00Z  2020  ...      4.0      8.0      3.0      2.0   
10  2020-04-19T00:00:00Z  2020  ...     10.0      5.0      5.0      NaN   
11  2020-04-26T00:00:00Z  2020  ...      2.0      7.0      4.0      9.0   
12  2020-04-19T00:00:00Z  2020  ...      2.0      2.0      2.0      1.0   

    hour_20  hour_21  hour_22  hour_23                               md5  \
0     338.0    295.0    297.0    172.0  6d91104f208db79b2e1cb6378737ee86   
1      47.0     26.0     27.0     17.0  6a7670fa4883bc4788f55ca5720e2e40   
3     144.0    113.0     62.0     44.0  c4c80f8e947790cfdbaf3f16f72048f7   
5       2.0      1.0      2.0      1.0  1c75a625642ddeda10ad96e41e38b57f   
7       8.0     15.0     16.0      4.0  f9901862c464533d69010b513236180f   
8       1.0      4.0      4.0      3.0  1bd2a5f07f89dee9e2a8b472103978c5   
9       6.0      3.0      2.0      3.0  5856ffa2eace86a8356933765bdb2273   
10      NaN      7.0      4.0      4.0  22cb7e009a3ee6f667d65188ba97a6bc   
11      7.0      5.0      8.0      2.0  5506b0f464ec0ab0b389fb875f1ce5d3   
12      4.0      4.0      2.0      1.0  c1e6c592ed1676b385a73b9de9f36315   

                       updated_on  
0   2020-05-04 05:28:51.591138+00  
1   2020-05-04 05:28:51.591138+00  
3   2020-05-04 05:28:51.591138+00  
5   2020-05-04 05:28:51.591138+00  
7   2020-05-04 05:28:51.591138+00  
8   2020-05-04 05:28:51.591138+00  
9   2020-05-04 05:28:51.591138+00  
10  2020-05-04 05:28:51.591138+00  
11  2020-05-04 05:28:51.591138+00  
12  2020-05-04 05:28:51.591138+00  

[10 rows x 41 columns]

In [ ]:
all_vehicle_traffic_volume.info()

<class 'pandas.core.frame.DataFrame'>
Index: 927941 entries, 0 to 1038414
Data columns (total 41 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   cartodb_id              927941 non-null  int64  
 1   the_geom                0 non-null       float64
 2   the_geom_webmercator    0 non-null       float64
 3   record_id               0 non-null       float64
 4   station_key             927941 non-null  int64  
 5   traffic_direction_seq   927941 non-null  int64  
 6   cardinal_direction_seq  927941 non-null  int64  
 7   classification_seq      927941 non-null  int64  
 8   date                    927941 non-null  object 
 9   year                    927941 non-null  int64  
 10  month                   927941 non-null  int64  
 11  day_of_week             927941 non-null  int64  
 12  public_holiday          927941 non-null  bool   
 13  school_holiday          927941 non-null  bool   
 14  daily_total             

In [ ]:
# Drop unecessary values
all_vehicle_traffic_volume = all_vehicle_traffic_volume.drop(columns =['the_geom','the_geom_webmercator','md5','updated_on','record_id'])

In [ ]:
ev_traffic_station = pd.read_csv('/content/drive/MyDrive/SIT764/Data/road_traffic_counts_station_reference.csv')
ev_traffic_station.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1783 entries, 0 to 1782
Data columns (total 42 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   the_geom                        0 non-null      float64
 1   cartodb_id                      1783 non-null   int64  
 2   the_geom_webmercator            0 non-null      float64
 3   record_id                       0 non-null      float64
 4   station_key                     1783 non-null   int64  
 5   station_id                      1783 non-null   object 
 6   name                            1783 non-null   object 
 7   road_name                       1783 non-null   object 
 8   full_name                       1783 non-null   object 
 9   common_road_name                1783 non-null   object 
 10  secondary_name                  1783 non-null   object 
 11  road_name_base                  1783 non-null   object 
 12  road_name_type                  17

In [ ]:
# Choose necessary values
ev_traffic_station = ev_traffic_station[['station_key','station_id','full_name','wgs84_latitude','wgs84_longitude','suburb','rms_region']]
ev_traffic_station.head(5)

station_key station_id                                     full_name  \
0     15828002       7211              Lily Lane, South of Fern Circuit   
1        58612      97340            Oxley Avenue, North of High Street   
2        56692      19035          Sydenham Road, West of Victoria Road   
3        58856      99931  Castlereagh Highway, South of Abattoirs Road   
4        56098      05622        Hanbury Street, East of Waratah Street   

   wgs84_latitude  wgs84_longitude        suburb  rms_region  
0      -32.940571       151.713120     Adamstown      Hunter  
1      -33.480766       145.534805      Hillston  South West  
2      -33.909630       151.162048  Marrickville      Sydney  
3      -32.559254       149.549698         Menah     Western  
4      -32.897934       151.735840      Mayfield      Hunter

In [ ]:
# Merging the data on station_key
traffic_volume_merger = all_vehicle_traffic_volume.merge(
    ev_traffic_station,
    left_on='station_key',
    right_on='station_key',
    how='left'
)

In [ ]:
traffic_volume_merger.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 927941 entries, 0 to 927940
Data columns (total 42 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   cartodb_id              927941 non-null  int64  
 1   station_key             927941 non-null  int64  
 2   traffic_direction_seq   927941 non-null  int64  
 3   cardinal_direction_seq  927941 non-null  int64  
 4   classification_seq      927941 non-null  int64  
 5   date                    927941 non-null  object 
 6   year                    927941 non-null  int64  
 7   month                   927941 non-null  int64  
 8   day_of_week             927941 non-null  int64  
 9   public_holiday          927941 non-null  bool   
 10  school_holiday          927941 non-null  bool   
 11  daily_total             927941 non-null  int64  
 12  hour_00                 807077 non-null  float64
 13  hour_01                 785407 non-null  float64
 14  hour_02             

In [ ]:
# Drop the missing value
traffic_volume_merger = traffic_volume_merger.dropna(subset=['wgs84_latitude'])
traffic_volume_merger.head(-5)


cartodb_id  station_key  traffic_direction_seq  \
0          3808503     15286009                      0   
1          3808504     15168302                      1   
2          3808506        57159                      0   
4          3808510     99990004                      0   
7          3808513        58868                      0   
...            ...          ...                    ...   
922537     9780146     15334006                      0   
922538     9780147     15286007                      1   
922539     9780148        56762                      0   
922546     9780155     15934006                      0   
922547     9780156     15648002                      0   

        cardinal_direction_seq  classification_seq                  date  \
0                            5                   2  2020-04-01T00:00:00Z   
1                            7                   2  2020-04-03T00:00:00Z   
2                            7                   2  2020-04-23T00:00:00Z   
4                            3                   3  2020-04-24T00:00:00Z   
7                            5                   3  2020-04-19T00:00:00Z   
...                        ...                 ...                   ...   
922537                       5                   2  2025-06-18T00:00:00Z   
922538                       7                   2  2025-06-04T00:00:00Z   
922539                       3                   3  2025-06-28T00:00:00Z   
922546                       3                   2  2025-06-03T00:00:00Z   
922547                       5                   3  2025-06-03T00:00:00Z   

        year  month  day_of_week  public_holiday  ...  hour_20  hour_21  \
0       2020      4            3           False  ...    338.0    295.0   
1       2020      4            5           False  ...     47.0     26.0   
2       2020      4            4           False  ...    144.0    113.0   
4       2020      4            5           False  ...      8.0     15.0   
7       2020      4            7           False  ...      NaN      7.0   
...      ...    ...          ...             ...  ...      ...      ...   
922537  2025      6            3           False  ...    345.0    329.0   
922538  2025      6            3           False  ...    751.0    626.0   
922539  2025      6            6           False  ...     31.0     14.0   
922546  2025      6            2           False  ...      NaN      NaN   
922547  2025      6            2           False  ...     25.0     21.0   

        hour_22  hour_23  station_id  \
0         297.0    172.0        7159   
1          27.0     17.0        6105   
2          62.0     44.0       57024   
4          16.0      4.0     7123-PR   
7           4.0      4.0      DNDSTC   
...         ...      ...         ...   
922537    283.0    115.0        9827   
922538    459.0    185.0        7150   
922539     13.0     16.0       23067   
922546      NaN      2.0       T0294   
922547      9.0     14.0        9849   

                                              full_name  wgs84_latitude  \
0              Cambridge Street, North of Joseph Street      -33.888157   
1        Great Western Highway, West of Curly Dick Road      -33.437302   
2                  Mona Vale Road, East of Addison Road      -33.686592   
4            Pages Road, North of Great Western Highway      -33.769588   
7       New England Highway, North of Severn River Road      -29.568588   
...                                                 ...             ...   
922537           Pittwater Road, South of Devitt Street      -33.719955   
922538         Showground Road, East of Victoria Avenue      -33.724922   
922539           Marsh Street, South of Rockwell Avenue      -33.934475   
922546   Great Western Highway, South of Carawatha Road      -33.615574   
922547         The Boulevarde, North of Torrington Road      -33.877975   

        wgs84_longitude          suburb  rms_region  
0            150.922974  Canley Heights      Sydney  
1      

In [ ]:
# Create necessay variable to identify which station that car go to
charger_coords = ev_charger_station_NSW[['Latitude', 'Longitude']]
sensor_coords = traffic_volume_merger[['station_id', 'wgs84_latitude', 'wgs84_longitude']].drop_duplicates()

In [ ]:
# Check for missing values
sensor_coords.isna().sum()

station_id         0
wgs84_latitude     0
wgs84_longitude    0
dtype: int64

In [ ]:
from sklearn.neighbors import NearestNeighbors
import numpy as np

# Sensor locations (latitude & longitude)
sensor_locations = sensor_coords[['wgs84_latitude', 'wgs84_longitude']].to_numpy()

# EV charger locations (latitude & longitude)
charger_locations = charger_coords.to_numpy()

# Use KNN to find the nearest sensor for each charger
knn = NearestNeighbors(n_neighbors=1, metric='euclidean')
knn.fit(sensor_locations)

# Find the index of the nearest sensor and the distance to it
distances, indices = knn.kneighbors(charger_locations)

# Assign the nearest sensor ID to each EV charging station
ev_charger_station_NSW['nearest_station_id'] = sensor_coords.iloc[indices.flatten()].station_id.values

# Convert distance from degrees to kilometers (1 degree ≈ 111 km)
ev_charger_station_NSW['distance_km'] = distances.flatten() * 111

In [ ]:
ev_charger_station_NSW[['Station name', 'nearest_station_id', 'distance_km']].head(10)


Station name nearest_station_id  distance_km
0                             Curlew Retreat              T0261   110.204597
1                           Donegal Farmstay              T0259    40.972922
2                           Glenrock Gardens              T0259    64.822818
3               Lightning Ridge Bowling Club              T0498   249.714874
4                                Baths Motel             BGBSTC   113.805858
5  Old Council Chambers Luxury Accommodation              T0259    59.400685
6            Econo Lodge Moree Spa Motor Inn             BGBSTC   113.288387
7                         Settlers Motor Inn              T0259    58.890747
8              Woodenbong Recreation Reserve              T0261    73.650685
9               Bonalbo Norman Johnston Park              T0261    49.607444

In [ ]:
# Merge data based on station ID
ev_chargers_linked = ev_charger_station_NSW.merge(
    traffic_volume_merger,
    left_on='nearest_station_id',
    right_on='station_id',
    how='left'
)

ev_chargers_linked.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3283910 entries, 0 to 3283909
Data columns (total 57 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   ObjId                   int64  
 1   Station name            object 
 2   Station address         object 
 3   Opening hours           object 
 4   Operator                object 
 5   Number of station       int64  
 6   Number of plugs         int64  
 7   Charger rating          object 
 8   Tesla                   int64  
 9   Type 2                  int64  
 10  J-1772                  int64  
 11  Latitude                float64
 12  Longitude               float64
 13  nearest_station_id      object 
 14  distance_km             float64
 15  cartodb_id              int64  
 16  station_key             int64  
 17  traffic_direction_seq   int64  
 18  cardinal_direction_seq  int64  
 19  classification_seq      int64  
 20  date                    object 
 21  year                    int64  

In [ ]:
# Drop unecessary values
ev_chargers_linked= ev_chargers_linked.drop(columns=['full_name','J-1772','wgs84_latitude','wgs84_longitude','cartodb_id'])

In [ ]:
print(ev_chargers_linked.columns)

Index(['ObjId', 'Station name', 'Station address', 'Opening hours', 'Operator',
       'Number of station', 'Number of plugs', 'Charger rating', 'Tesla',
       'Type 2', 'Latitude', 'Longitude', 'nearest_station_id', 'distance_km',
       'station_key', 'traffic_direction_seq', 'cardinal_direction_seq',
       'classification_seq', 'date', 'year', 'month', 'day_of_week',
       'public_holiday', 'school_holiday', 'daily_total', 'hour_00', 'hour_01',
       'hour_02', 'hour_03', 'hour_04', 'hour_05', 'hour_06', 'hour_07',
       'hour_08', 'hour_09', 'hour_10', 'hour_11', 'hour_12', 'hour_13',
       'hour_14', 'hour_15', 'hour_16', 'hour_17', 'hour_18', 'hour_19',
       'hour_20', 'hour_21', 'hour_22', 'hour_23', 'station_id', 'suburb',
       'rms_region'],
      dtype='object')


In [ ]:
# Print to check the year
ev_chargers_linked['year'].unique()

array([2020, 2021, 2022, 2023, 2024, 2025])

In [ ]:
# Ensure date column is datetime
ev_chargers_linked['date'] = pd.to_datetime(ev_chargers_linked['date'])

# Add month and year columns just in case they're not parsed correctly
ev_chargers_linked['month'] = ev_chargers_linked['date'].dt.month
ev_chargers_linked['year'] = ev_chargers_linked['date'].dt.year

**3. Web scraping to get the weather data**

In [6]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

web_url = "https://www.timeanddate.com/weather/australia/sydney/historic?month=1&year=2020"
resp = requests.get(web_url)
resp.raise_for_status()

# 2. Parse HTML bằng BeautifulSoup
soup = BeautifulSoup(resp.text, "html.parser")
print(soup.prettify())

<!DOCTYPE html>
<!--
scripts and programs that download content transparent to the user are not allowed without permission
-->
<html lang="en">
 <head>
  <meta content="text/html; charset=utf-8" http-equiv="Content-Type"/>
  <title>
   Weather in January 2020 in Sydney, New South Wales, Australia
  </title>
  <meta content="Weather reports from January 2020 in Sydney, New South Wales, Australia with highs and lows" name="description"/>
  <meta content="max-image-preview:large" name="robots"/>
  <meta content="https://www.timeanddate.com/scripts/cityog.php?title=Past%20Weather%20in&amp;tint=0x007b7a&amp;city=Sydney&amp;state=New%20South%20Wales&amp;country=Australia&amp;image=sydney1" property="og:image"/>
  <meta content="1366" property="og:image:width"/>
  <meta content="738" property="og:image:height"/>
  <meta content="website" property="og:type"/>
  <style>
   @font-face{font-family:iconfont;src:url("/common/fonts/iconfont.woff2?v8") format("woff2"),url("/common/fonts/iconfont.woff

In [7]:
BASE_URL = "https://www.timeanddate.com/weather/australia/sydney/historic"

def get_hd_list(month:int, year:int):
  resp = requests.get(BASE_URL, params={"month": month, "year": year})
  resp.raise_for_status()
  soup = BeautifulSoup(resp.text, "html.parser")
  return [
      opt["value"]
      for opt in soup.select("select#wt-his-select option")
  ]

def parse_table(html: str):
    soup = BeautifulSoup(html, "html.parser")
    tbl = soup.find("table", id="wt-his")
    if not tbl:
        return pd.DataFrame()

    thead_rows = tbl.thead.find_all("tr")
    header_tr = thead_rows[-1]
    headers = [
        th.get_text(strip=True)
        for th in header_tr.find_all("th")
    ]


    rows = []
    for tr in tbl.tbody.find_all("tr"):
        cells = [
            cell.get_text(" ", strip=True)
            for cell in tr.find_all(["th","td"])
        ]

        if len(cells) == len(headers):
            rows.append(cells)

    return pd.DataFrame(rows, columns=headers)

def scrape_month(month:int, year:int):
  try:
    hds = get_hd_list(month, year)
  except Exception as e:
      print(f"Failed to fetch date list: {e}")
      return pd.DataFrame()

  dfs = []
  for hd in hds:
      r = requests.get(BASE_URL, params={"month": month, "year": year, "hd": hd})
      r.raise_for_status()
      df = parse_table(r.text)
      if not df.empty:
           # add Date column from hd
          df.insert(0, "Date", pd.to_datetime(hd, format="%Y%m%d"))
          dfs.append(df)
  return pd.concat(dfs, ignore_index=True) if dfs else pd.DataFrame()


# Example: get the entire table for February 2020
jan_2020_df = scrape_month(1, 2020)
jan_2020_df.head(15)


Date                 Time     Temp          Weather    Wind     \
0  2020-01-01  12:00 am Wed, Jan 1    68 °F   Partly cloudy.  22 mph  ↑   
1  2020-01-01             12:30 am    68 °F   Partly cloudy.  18 mph  ↑   
2  2020-01-01              1:00 am    68 °F   Partly cloudy.  16 mph  ↑   
3  2020-01-01              1:30 am    66 °F   Partly cloudy.  17 mph  ↑   
4  2020-01-01              2:00 am    68 °F  Passing clouds.  17 mph  ↑   
5  2020-01-01              2:30 am    66 °F  Passing clouds.  18 mph  ↑   
6  2020-01-01              3:00 am    66 °F  Passing clouds.  17 mph  ↑   
7  2020-01-01              3:30 am    66 °F  Passing clouds.  17 mph  ↑   
8  2020-01-01              4:00 am    68 °F  Passing clouds.  17 mph  ↑   
9  2020-01-01              4:30 am    68 °F  Passing clouds.  17 mph  ↑   
10 2020-01-01              5:00 am    68 °F  Passing clouds.  16 mph  ↑   
11 2020-01-01              5:30 am    68 °F  Passing clouds.  17 mph  ↑   
12 2020-01-01              6:00 am    66 °F    Partly sunny.  17 mph  ↑   
13 2020-01-01              6:30 am    68 °F   Broken clouds.  16 mph  ↑   
14 2020-01-01              7:00 am    68 °F   Broken clouds.  15 mph  ↑   

   Humidity  Barometer Visibility  
0       73%  29.95 "Hg        N/A  
1       73%  29.92 "Hg        N/A  
2       78%  29.92 "Hg        N/A  
3       78%  29.92 "Hg        N/A  
4       78%  29.92 "Hg        N/A  
5       78%  29.89 "Hg        N/A  
6       78%  29.89 "Hg        N/A  
7       73%  29.89 "Hg        N/A  
8       73%  29.89 "Hg        N/A  
9       73%  29.89 "Hg        N/A  
10      69%  29.89 "Hg        N/A  
11      73%  29.92 "Hg        N/A  
12      73%  29.92 "Hg        N/A  
13      69%  29.92 "Hg        N/A  
14      69%  29.95 "Hg        N/A

In [10]:
import pandas as pd

all_dfs = []

start_year, start_month = 2020, 1
end_year, end_month     = 2025, 6

for year in range(start_year, end_year + 1):
    # For the first year, start from start_month; for the last year, end at end_month
    m_start = start_month if year == start_year else 1
    m_end   = end_month   if year == end_year   else 12

    for month in range(m_start, m_end + 1):
        df_month = scrape_month(month, year)
        if not df_month.empty:
            all_dfs.append(df_month)
        else:
            print("no data")

# Combine everything
if all_dfs:
    full_df = pd.concat(all_dfs, ignore_index=True)
    print(f"\nFinished! Total number of records: {len(full_df)}")
else:
    full_df = pd.DataFrame()
    print("No data was scraped!")



Finished! Total number of records: 90701


In [ ]:
full_df.to_csv("sydney_2020_2025.csv", index=False)

In [ ]:
# Read the weather data
weather_data = pd.read_csv('/content/drive/MyDrive/SIT764/Data/sydney_2020_2025.csv')
weather_data.head(5)

Date              Time  Unnamed: 2   Temp          Weather     Wind  \
0  2020-01-01  00:00 Wed, 1 Jan         NaN  20 °C   Partly cloudy.  35 km/h   
1  2020-01-01             00:30         NaN  20 °C   Partly cloudy.  30 km/h   
2  2020-01-01             01:00         NaN  20 °C   Partly cloudy.  26 km/h   
3  2020-01-01             01:30         NaN  19 °C   Partly cloudy.  28 km/h   
4  2020-01-01             02:00         NaN  20 °C  Passing clouds.  28 km/h   

  Unnamed: 6 Humidity  Barometer Visibility  
0          ↑      73%  1014 mbar        NaN  
1          ↑      73%  1013 mbar        NaN  
2          ↑      78%  1013 mbar        NaN  
3          ↑      78%  1013 mbar        NaN  
4          ↑      78%  1013 mbar        NaN

In [ ]:
# Create the function to calculate the mean of weather
def compute_weather_average(dataframe):
    df = dataframe.copy()

    df['Temp_numeric'] = (
            df['Temp']
              .str.replace(r"[^\d\.]", "", regex=True)
              .astype(float)
        )

    # Parse humidity: remove “%”
    df['Humidity_numeric'] = (
        df['Humidity']
          .str.replace('%', '', regex=False)
          .astype(float)
    )

    # Parse wind: extract the numeric part
    df['Wind_numeric'] = (
        df['Wind']
          .str.extract(r'(\d+\.?\d*)')[0]
          .astype(float)
    )

    # Group by Date and compute means
    daily_avg = (
        df
        .groupby('Date')
        .agg(
            Avg_Temp     = ('Temp_numeric',    'mean'),
            Avg_Humidity = ('Humidity_numeric','mean'),
            Avg_Wind     = ('Wind_numeric',    'mean'),
        )
        .round(2)
        .reset_index()
    )

    return daily_avg

In [ ]:
# Implement the function
weather_data_sydney = compute_weather_average(weather_data)
weather_data_sydney.head(15)

Date  Avg_Temp  Avg_Humidity  Avg_Wind
0   2020-01-01     21.55         72.00     24.98
1   2020-01-02     22.43         70.00     20.96
2   2020-01-03     24.80         73.50     18.02
3   2020-01-04     30.10         54.04     27.47
4   2020-01-05     22.28         66.84     34.90
5   2020-01-06     21.92         70.82     16.89
6   2020-01-07     24.31         78.51     19.02
7   2020-01-08     22.42         84.46     30.98
8   2020-01-09     22.25         82.60     23.35
9   2020-01-10     26.24         69.48     25.78
10  2020-01-11     21.31         73.22     36.40
11  2020-01-12     20.22         72.15     23.37
12  2020-01-13     21.92         67.85     16.28
13  2020-01-14     23.21         71.19     20.44
14  2020-01-15     24.04         71.30     22.93

In [ ]:
# Ensure the date columns are datetime
# Convert 'Date' column to datetime (if it isn't already)
weather_data_sydney['Date'] = pd.to_datetime(weather_data_sydney['Date'], errors='coerce')

# Ensure the date columns are datetime and localize to UTC
weather_data_sydney['date'] = weather_data_sydney['Date'].dt.tz_localize('UTC')
# Merge on the date column
#    By default, this is an inner join: only dates present in both will remain
merged = pd.merge(
    ev_chargers_linked,
    weather_data_sydney,
    on='date',
    how='inner',
    suffixes=('_h','_duy')
)

# Inspect the result
print(merged.head())


   ObjId    Station name                                   Station address  \
0      1  Curlew Retreat  399 Wylie Creek Rd, Maryland NSW 2372, Australia   
1      1  Curlew Retreat  399 Wylie Creek Rd, Maryland NSW 2372, Australia   
2      1  Curlew Retreat  399 Wylie Creek Rd, Maryland NSW 2372, Australia   
3      1  Curlew Retreat  399 Wylie Creek Rd, Maryland NSW 2372, Australia   
4      1  Curlew Retreat  399 Wylie Creek Rd, Maryland NSW 2372, Australia   

  Opening hours Operator  Number of station  Number of plugs Charger rating  \
0     Open 24/7      NaN                  1                1             kW   
1     Open 24/7      NaN                  1                1             kW   
2     Open 24/7      NaN                  1                1             kW   
3     Open 24/7      NaN                  1                1             kW   
4     Open 24/7      NaN                  1                1             kW   

   Tesla  Type 2  ...  hour_21  hour_22 hour_23  station

In [ ]:
# Save the dataset
merged.to_csv('merge_data.csv', index=False)